### PR TITLE
Implemented Policy update

### PR DESCRIPTION
**Pull request by samuel4x4**
Implemented Policy update

### CODE REVIEW

**Comment by samuel4x4** [comment]
@jamiehannaford, @haphan: I have no idea why the CI check failed, can you help me please? The only change I did in composer.json was to put temporary, in my form, the package name as "samuel4x4/openstack", so I can install it with Composer and test it.

File: composer.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-    "name": "php-opencloud/openstack",
+    "name": "samuel4x4/openstack",
```
Comment (jamiehannaford) [inline composer.json]: this shouldn't be here

File: src/Common/Api/OperatorTrait.php [hunk 2]
```diff
@@ -106,8 +105,6 @@ public function getOperation(array $definition): Operation
      */
     protected function sendRequest(Operation $operation, array $userValues = [], bool $async = false)
     {
-        $operation->validate($userValues);
```
Comment (jamiehannaford) [inline src/Common/Api/OperatorTrait.php]: why are we not validating?

File: src/Common/Resource/Updateable.php [hunk 3]
```diff
@@ -12,7 +12,8 @@
     /**
      * Update the current resource with the configuration set out in the user options.
      *
-     * @return void
+     * @param array $userOptions
+     * @return self
      */
-    public function update();
+    public function update(array $userOptions): Updateable;
```
Comment (jamiehannaford) [inline src/Common/Resource/Updateable.php]: you're changing a public API. we cannot do this without a major version bump.

File: src/Identity/v3/Models/Credential.php [hunk 4]
```diff
@@ -39,10 +39,13 @@ class Credential extends OperatorResource implements Creatable, Updateable, Retr
 
     /**
      * {@inheritDoc}
+     *
+     * @param array $data {@see \OpenStack\Identity\v3\Api::postCredentials}
      */
     public function create(array $data): Creatable
     {
         $response = $this->execute($this->api->postCredentials(), $data);
+
```
Comment (jamiehannaford) [inline src/Identity/v3/Models/Credential.php]: please remove

File: src/Identity/v3/Models/Credential.php [hunk 5]
```diff
@@ -52,15 +55,19 @@ public function create(array $data): Creatable
     public function retrieve()
     {
         $response = $this->executeWithState($this->api->getCredential());
+
```
Comment (jamiehannaford) [inline src/Identity/v3/Models/Credential.php]: please remove

File: src/Identity/v3/Models/Credential.php [hunk 6]
```diff
@@ -52,15 +55,19 @@ public function create(array $data): Creatable
...
     /**
      * {@inheritDoc}
+     *
+     * @param array $data {@see \OpenStack\Identity\v3\Api::patchCredential}
      */
-    public function update()
+    public function update(array $data): Updateable
     {
         $response = $this->executeWithState($this->api->patchCredential());
+
```
Comment (jamiehannaford) [inline src/Identity/v3/Models/Credential.php]: please remove here and all other places

File: src/Identity/v3/Service.php [hunk 7]
```diff
@@ -423,6 +423,21 @@ public function createPolicy(array $options): Models\Policy
...
 
     /**
+     * Updates a policy object according to the provided options.
+     *
+     * @param string $id
+     * @param array $options {@see \OpenStack\Identity\v3\Api::patchPolicy}
+     *
+     * @return Models\Policy
+     */
+    public function updatePolicy(string $id, array $options): Models\Policy
```
Comment (jamiehannaford) [inline src/Identity/v3/Service.php]: this is not how operations are done. the user retrieves a model then performs the update on that object.

File: composer.json (hunk 1 above)
Comment (samuel4x4) [reply composer.json]: you're right

File: src/Common/Api/OperatorTrait.php (hunk 2 above)
Comment (samuel4x4) [reply src/Common/Api/OperatorTrait.php]: it was removed by mistake

**Comment by samuel4x4** [comment]
@jamiehannaford thank you for your review. I was wrong, I didn't notice there's already an existing possibility to update a Policy. Even that, I rushed and made mistakes. I'll close the PR and I will gladly use the existing feature.

**Pull request closed without merging**